File: JASSjr_search.py
#!/usr/bin/env python3

# JASSjr_search.py
# Copyright (c) 2023 Vaughan Kitchen
# Minimalistic BM25 search engine.

from array import array
from collections import deque
import math
import struct
import sys

k1 = 0.9 # BM25 k1 parameter
b = 0.4 # BM25 b parameter

def read_file(filename):
    with open(filename, mode='rb') as file:
            return file.read()

def read_lines(filename):
    with open(filename) as file:
            return file.readlines()

# decode the vocabulary (one byte length, string, '\0', 4 byte where, 4 byte size)
def decode_vocab(buffer):
    offset = 0
    while offset < len(buffer):
        length, = struct.unpack_from('B', buffer, offset=offset)
        offset += 1

        word, = struct.unpack_from(f'{length}s', buffer, offset=offset)
        offset += length + 1 # Null terminated

        where, size = struct.unpack_from('ii', buffer, offset=offset)
        offset += 8

        yield word.decode(), where, size

contents_vocab = read_file('vocab.bin')
contents_postings = read_file('postings.bin')
doc_lengths = array('i', read_file('lengths.bin')) # Read the document lengths
doc_ids = read_lines('docids.bin') # Read the primary_keys

# Compute the average document length for BM25
average_length = sum(doc_lengths) / len(doc_lengths)
vocab = {}

# Build the vocabulary in memory
for word, offset, size in decode_vocab(contents_vocab):
    vocab[word] = (offset, size)

# Search (one query per line)
for query in sys.stdin:
    query_id = 0
    accumulators = [(0, 0)] * len(doc_lengths) # array of rsv values

    # If the first token is a number then assume a TREC query number, and skip it
    terms = deque(query.split())
    if terms[0].isnumeric():
        query_id = terms.popleft()

    for term in terms:
        # Does the term exist in the collection?
        try:
            offset, size = vocab[term]

            postings_length = size / 8
            # Compute the IDF component of BM25 as log(N/n).
            # if IDF == 0 then don't process this postings list as the BM25 contribution of this term will be zero.
            idf = math.log(len(doc_lengths) / (postings_length))

            # Seek and read the postings list
            for docid, freq in struct.iter_unpack('ii', contents_postings[offset:offset+size]):
                # Process the postings list by simply adding the BM25 component for this document into the accumulators array
                rsv = idf * ((freq * (k1 + 1)) / (freq + k1 * (1 - b + b * (doc_lengths[docid] / average_length))))
                current_rsv = accumulators[docid][0]
                accumulators[docid] = (current_rsv + rsv, docid)
        except KeyError:
            pass

    # Sort the results list. Tie break on the document ID.
    accumulators.sort(reverse=True)

    # Print the (at most) top 1000 documents in the results list in TREC eval format which is:
    # query-id Q0 document-id rank score run-name
    for i, (rsv, docid) in enumerate(accumulators, start=1):
        if rsv == 0 or i == 1001:
            break
        print("{} Q0 {} {} {:.4f} JASSjr".format(query_id, doc_ids[docid][:-1], i, rsv))

File: JASSjr_search.js
#!/usr/bin/env node

// JASSjr_search.js
// Copyright (c) 2023, 2024 Vaughan Kitchen
// Minimalistic BM25 search engine.

var fs = require('fs');
var os = require('os');
var readline = require('readline').createInterface({
	input: process.stdin,
	output: process.stdout,
});

var k1 = 0.9; // BM25 k1 parameter
var b = 0.4; // BM25 b parameter

var vocab_raw = fs.readFileSync('vocab.bin');
var doc_lengths = fs.readFileSync('lengths.bin'); // Read the document lengths
var ids = fs.readFileSync('docids.bin', 'utf8').split(os.EOL); // Read the primary_keys

// Compute the average document length for BM25
var documents_in_collection = doc_lengths.length / 4;
var average_document_length = 0;
for (var i = 0; i < doc_lengths.length; i += 4)
	average_document_length += doc_lengths.subarray(i, i+4).readUInt32LE();
average_document_length /= documents_in_collection;

var vocab = {};

// Build the vocabulary in memory (one byte length, string, '\0', 4 byte where, 4 byte size)
for (var offset = 0; offset < vocab_raw.length;) {
	var length = vocab_raw[offset];
	offset++;

	var word = vocab_raw.subarray(offset, offset+length).toString();
	offset += length+1; // null terminated

	var where = vocab_raw.subarray(offset, offset+4).readInt32LE();
	var size = vocab_raw.subarray(offset+4, offset+8).readInt32LE();
	offset += 8;

	vocab[word] = [where, size];
}

// Open the postings list file
var postings = new Int32Array(documents_in_collection * 2);
var postings_fh = fs.openSync('postings.bin');

// Search (one query per line)
readline.question('', search);

function search(query) {
	var terms = query.split(' ');

	var query_id = 0;
	var accumulators = new Array(documents_in_collection); // array of rsv values

	// If the first token is a number then assume a TREC query number, and skip it
	if (!isNaN(terms[0]))
		query_id = terms.shift();

	terms.forEach(function (term) {
		// Does the term exist in the collection?
		var pair = vocab[term];
		if (pair === undefined)
			return;

		var [where, size] = pair;
		var documents_in_postings = size / 8;

		// if IDF == 0 then don't process this postings list as the BM25 contribution of this term will be zero.
		if (documents_in_collection == documents_in_postings)
			return;

		// Seek and read the postings list
		fs.readSync(postings_fh, postings, 0, size, where);

		// Compute the IDF component of BM25 as log(N/n).
		var idf = Math.log(documents_in_collection / documents_in_postings)

		for (var i = 0; i < documents_in_postings * 2; i += 2) {
			var docid = postings[i];
			var freq = postings[i+1];

			var doc_length = doc_lengths.subarray(docid*4, docid*4+4).readUInt32LE();

			// Process the postings list by simply adding the BM25 component for this document into the accumulators array
			var rsv = idf * ((freq * (k1 + 1)) / (freq + k1 * (1 - b + b * (doc_length / average_document_length))))
			var current = accumulators[docid] || [0, 0]
			accumulators[docid] = [docid, current[1] + rsv];
		}
	});

	// Sort the results list. Tie break on the document ID.
	accumulators.sort(function (a, b) { return b[1] - a[1] || b[0] - a[0] });

	// Print the (at most) top 1000 documents in the results list in TREC eval format which is:
	// query-id Q0 document-id rank score run-name
	for (var i = 0; i < accumulators.length; i++) {
		var pair = accumulators[i];
		if (!pair || i == 1000)
			break;
		var docid = pair[0];
		var rsv = pair[1];
		console.log('%d Q0 %s %d %s JASSjr', query_id, ids[docid], i+1, rsv.toLocaleString('en-US', { minimumFractionDigits: 4, maximumFractionDigits: 4 }));
	}

	readline.question('', search);
}
